File: test/cpu/cas2.cc
#define BOOST_TEST_DYN_LINK
#include "68040.hpp"
#include "test.hpp"

#include <boost/test/data/monomorphic.hpp>
#include <boost/test/data/test_case.hpp>
#include <boost/test/unit_test.hpp>
namespace bdata = boost::unit_test::data;
struct F_CAS2 {
    F_CAS2() {
        // CAS2.W %D4:%D7, %D3:%D6, (%A2):(%A5)
        TEST::SET_W(0, 0006374);
        TEST::SET_W(2, 1 << 15 | 2 << 12 | 3 << 6 | 4);
        TEST::SET_W(4, 1 << 15 | 5 << 12 | 6 << 6 | 7);
        TEST::SET_W(6, TEST_BREAK);

        // CAS2.L %D4:%D7, %D3:%D6, (%A2):(%A5)
        TEST::SET_W(8, 0007374);
        TEST::SET_W(10, 1 << 15 | 2 << 12 | 3 << 6 | 4);
        TEST::SET_W(12, 1 << 15 | 5 << 12 | 6 << 6 | 7);
        TEST::SET_W(14, TEST_BREAK);

        jit_compile(0, 18);
    }
};
BOOST_FIXTURE_TEST_SUITE(CAS2, Prepare, *boost::unit_test::fixture<F_CAS2>())


BOOST_AUTO_TEST_SUITE(Word)

BOOST_AUTO_TEST_CASE(T) {
    TEST::SET_W(0x1000, 0x2222);
    TEST::SET_W(0x1010, 0x3333);
    cpu.A[2] = 0x1000;
    cpu.A[5] = 0x1010;
    cpu.D[3] = 0x3030;
    cpu.D[4] = 0x2222;
    cpu.D[6] = 0x4040;
    cpu.D[7] = 0x3333;
    run_test(0);
    BOOST_TEST(TEST::GET_W(0x1000) == 0x3030);
    BOOST_TEST(TEST::GET_W(0x1010) == 0x4040);
}
BOOST_AUTO_TEST_CASE(F1) {
    TEST::SET_W(0x1000, 0x2222);
    TEST::SET_W(0x1010, 0x3333);
    cpu.A[2] = 0x1000;
    cpu.A[5] = 0x1010;
    cpu.D[3] = 0x3030;
    cpu.D[4] = 0x2227;
    cpu.D[6] = 0x4040;
    cpu.D[7] = 0x3333;
    run_test(0);
    BOOST_TEST(cpu.D[4] == 0x2222);
    BOOST_TEST(cpu.D[7] == 0x3333);
}

BOOST_AUTO_TEST_CASE(F2) {
    TEST::SET_W(0x1000, 0x2222);
    TEST::SET_W(0x1010, 0x3333);
    cpu.A[2] = 0x1000;
    cpu.A[5] = 0x1010;
    cpu.D[3] = 0x3030;
    cpu.D[4] = 0x2222;
    cpu.D[6] = 0x4040;
    cpu.D[7] = 0x3339;
    run_test(0);
    BOOST_TEST(cpu.D[4] == 0x2222);
    BOOST_TEST(cpu.D[7] == 0x3333);
}

BOOST_AUTO_TEST_CASE(traced) {
    TEST::SET_W(0x1000, 0x2222);
    TEST::SET_W(0x1010, 0x3333);
    cpu.T = 1;
	run_test(0);
	BOOST_TEST(ex_n == EXCEPTION_NUMBER::TRACE );
}

BOOST_AUTO_TEST_SUITE_END()

BOOST_AUTO_TEST_SUITE(Long)

BOOST_AUTO_TEST_CASE(T) {
    TEST::SET_L(0x1000, 0x22222222);
    TEST::SET_L(0x1010, 0x33333333);
    cpu.A[2] = 0x1000;
    cpu.A[5] = 0x1010;
    cpu.D[3] = 0x30303030;
    cpu.D[4] = 0x22222222;
    cpu.D[6] = 0x40404040;
    cpu.D[7] = 0x33333333;
    run_test(8);
    BOOST_TEST(TEST::GET_L(0x1000) == 0x30303030);
    BOOST_TEST(TEST::GET_L(0x1010) == 0x40404040);
}

BOOST_AUTO_TEST_CASE(F1) {
    TEST::SET_L(0x1000, 0x22222222);
    TEST::SET_L(0x1010, 0x33333333);
    cpu.A[2] = 0x1000;
    cpu.A[5] = 0x1010;
    cpu.D[3] = 0x30303030;
    cpu.D[4] = 0x22222227;
    cpu.D[6] = 0x40404040;
    cpu.D[7] = 0x33333333;
    run_test(8);
    BOOST_TEST(cpu.D[4] == 0x22222222);
    BOOST_TEST(cpu.D[7] == 0x33333333);
}

BOOST_AUTO_TEST_CASE(F2) {
    TEST::SET_L(0x1000, 0x22222222);
    TEST::SET_L(0x1010, 0x33333333);
    cpu.A[2] = 0x1000;
    cpu.A[5] = 0x1010;
    cpu.D[3] = 0x30303030;
    cpu.D[4] = 0x22222222;
    cpu.D[6] = 0x40404040;
    cpu.D[7] = 0x33333339;
    run_test(8);
    BOOST_TEST(cpu.D[4] == 0x22222222);
    BOOST_TEST(cpu.D[7] == 0x33333333);
}

BOOST_AUTO_TEST_CASE(traced) {
    TEST::SET_L(0x1000, 0x22222222);
    TEST::SET_L(0x1010, 0x33333333);
    cpu.T = 1;
	run_test(8);
	BOOST_TEST(ex_n == EXCEPTION_NUMBER::TRACE );
}
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE_END()

File: test/lua/cpu/cas2.lua
-- CAS2.W %D3:%D4, %D5:%D6, (%A1):(%A2)
MEM[0].W = OCT"0006374"
MEM[2].W = 1 << 15 | 1 << 12 | 5 << 6 | 3
MEM[4].W = 1 << 15 | 2 << 12 | 6 << 6 | 4
MEM[6].W = TEST.BREAK

-- CAS2.L %D4:%D5, %D6:%D7, (%A2):(%A3)
MEM[8].W = OCT"0007374"
MEM[10].W = 1 << 15 | 2 << 12 | 6 << 6 | 4
MEM[12].W = 1 << 15 | 3 << 12 | 7 << 6 | 5
MEM[14].W = TEST.BREAK

JIT_COMPILE(0, 16)

local run = {}
run.W = {
    t = function()
        MEM[0x1000].W = 0x2222
        MEM[0x1010].W = 0x3333
        CPU.A[2] = 0x1000
        CPU.A[3] = 0x1010
        CPU.D[4].W = 0x2222
        CPU.D[5].W = 0x3333
        CPU.D[6].W = 0x3030
        CPU.D[7].W = 0x4040
        TEST.RUN(0)
        TEST.CHECK("MEM[0x1000].W", 0x3030)
        TEST.CHECK("MEM[0x1010].W", 0x4040)
    end,
    f1 = function()
        MEM[0x1000].W = 0x2222
        MEM[0x1010].W = 0x3333
        CPU.A[2] = 0x1000
        CPU.A[3] = 0x1010
        CPU.D[4].W = 0x2227
        CPU.D[5].W = 0x3030
        CPU.D[6].W = 0x3333
        CPU.D[7].W = 0x4040
        TEST.RUN(0)
        TEST.CHECK("CPU.D[4].W", 0x2222)
        TEST.CHECK("CPU.D[5].W", 0x3333)
    end,
    f2 = function()
        MEM[0x1000].W = 0x2222
        MEM[0x1010].W = 0x3333
        CPU.A[2] = 0x1000
        CPU.A[3] = 0x1010
        CPU.D[4].W = 0x2222
        CPU.D[5].W = 0x3030
        CPU.D[6].W = 0x3339
        CPU.D[7].W = 0x4040
        TEST.RUN(0)
        TEST.CHECK("CPU.D[4].W", 0x2222)
        TEST.CHECK("CPU.D[5].W", 0x3333)
    end,
    trace = function()
        MEM[0x1000].W = 0x2222
        MEM[0x1010].W = 0x3333
        CPU.A[2] = 0x1000
        CPU.A[3] = 0x1010
        trace_test(0)
   end,
}

run.L = {
    t = function()
        MEM[0x1000].L = 0x22222222
        MEM[0x1010].L = 0x33333333
        CPU.A[3] = 0x1000
        CPU.A[4] = 0x1010
        CPU.D[5].L = 0x22222222
        CPU.D[6].L = 0x33333333
        CPU.D[7].L = 0x30303030
        CPU.D[8].L = 0x40404040
        TEST.RUN(8)
        TEST.CHECK("MEM[0x1000].L", 0x30303030)
        TEST.CHECK("MEM[0x1010].L", 0x40404040)
    end,
    f1 = function()
        MEM[0x1000].L = 0x22222222
        MEM[0x1010].L = 0x33333333
        CPU.A[3] = 0x1000
        CPU.A[4] = 0x1010
        CPU.D[5].L = 0x22222227
        CPU.D[6].L = 0x30303030
        CPU.D[7].L = 0x33333333
        CPU.D[8].L = 0x40404040
        TEST.RUN(8)
        TEST.CHECK("CPU.D[5].L", 0x22222222)
        TEST.CHECK("CPU.D[6].L", 0x33333333)
    end,
    f2 = function()
        MEM[0x1000].L = 0x22222222
        MEM[0x1010].L = 0x33333333
        CPU.A[3] = 0x1000
        CPU.A[4] = 0x1010
        CPU.D[5].L = 0x22222222
        CPU.D[6].L = 0x30303030
        CPU.D[7].L = 0x33333339
        CPU.D[8].L = 0x40404040
        TEST.RUN(8)
        TEST.CHECK("CPU.D[5].L", 0x22222222)
        TEST.CHECK("CPU.D[6].L", 0x33333333)
    end,
    trace = function()
        MEM[0x1000].L = 0x22222222
        MEM[0x1010].L = 0x33333333
        CPU.A[3] = 0x1000
        CPU.A[4] = 0x1010
        trace_test(8)
   end,
}

return run